File: src/2025-03-04-03-28-40--deepseek-chat-0304/go/exercises/practice/variable-length-quantity/variable_length_quantity.go
package variablelengthquantity

import "errors"

func EncodeVarint(input []uint32) []byte {
	var result []byte
	for _, n := range input {
		// Handle zero case
		if n == 0 {
			result = append(result, 0)
			continue
		}

		var bytes []byte
		for n > 0 {
			// Take the lowest 7 bits
			byteVal := byte(n & 0x7F)
			n >>= 7
			// Set the continuation bit if there are more bytes
			if len(bytes) > 0 {
				byteVal |= 0x80
			}
			bytes = append([]byte{byteVal}, bytes...)
		}
		result = append(result, bytes...)
	}
	return result
}

func DecodeVarint(input []byte) ([]uint32, error) {
	var result []uint32
	var current uint32

	for i := 0; i < len(input); {
		b := input[i]
		// Add the 7 bits to the current value
		current = (current << 7) | uint32(b&0x7F)
		
		// If continuation bit is set, continue to next byte
		if b&0x80 != 0 {
			i++
			continue
		}
		
		// No continuation bit, this is the last byte of the number
		result = append(result, current)
		current = 0
		i++
	}

	return result, nil
}

File: src/2025-03-04-03-28-40--deepseek-chat-0304/java/exercises/practice/variable-length-quantity/src/main/java/VariableLengthQuantity.java
import java.util.List;
import java.util.ArrayList;
import java.util.Collections;

class VariableLengthQuantity {

    List<String> encode(List<Integer> numbers) {
        List<String> result = new ArrayList<>();
        for (int number : numbers) {
            List<String> encodedBytes = new ArrayList<>();
            if (number == 0) {
                encodedBytes.add("00");
            } else {
                while (number > 0) {
                    byte b = (byte)(number & 0x7F);
                    number >>>= 7;
                    if (!encodedBytes.isEmpty()) {
                        b |= 0x80; // Set continuation bit
                    }
                    encodedBytes.add(String.format("%02X", b));
                }
                Collections.reverse(encodedBytes);
            }
            result.addAll(encodedBytes);
        }
        return result;
    }

    List<String> decode(List<Integer> bytes) {
        List<String> result = new ArrayList<>();
        int number = 0;
        for (int b : bytes) {
            number = (number << 7) | (b & 0x7F);
            if ((b & 0x80) == 0) {
                result.add(String.format("%02X", number));
                number = 0;
            }
        }
        return result;
    }
}
